### ISSUE TITLE
Workflow for examples in testing chapter is unclear

### ISSUE THREAD

**Issue by yminsky** [realworldocaml/book#3248]
I'm not sure how to test individual examples in the testing chapter in a reasonable way. If I switch to an individual directory and run "dune runtest", it fails, like this:

```
[yminsky@viserion simple_expect_test]$ dune runtest
File "dune", line 3, characters 17-22:
3 |  (libraries base stdio)
                     ^^^^^
Error: Library "stdio" not found.
Hint: try: dune external-lib-deps --missing @runtest
```

The code in question can be found [here](https://github.com/realworldocaml/book/tree/testing-chapter/book/testing/examples/multi_block_expect_test).

I can invoke this just fine from the README, with a stanza like this:

    ```sh dir=examples/soup_test
      $ dune runtest
    ```

All of this is while I have the "right" opam setup, basically, using opam installed as a local switch.  But if I switch to my "regular" opam environment, I can get things working from the command line.

```
[yminsky@viserion soup_test]$ opam switch 4.09.0
# Run eval $(opam env --switch=4.09.0) to update the current shell environment
[yminsky@viserion soup_test]$ eval $(opam env --switch=4.09.0)
[NOTE] To make opam select the switch 4.09.0 in the current shell, add --set-switch or set OPAMSWITCH
[yminsky@viserion soup_test]$ dune runtest
     patdiff (internal) (exit 1)
(cd _build/default && /home/yminsky/bin/patdiff -keep-whitespace -location-style omake -unrefined test.ml test.ml.corrected)
------ test.ml
++++++ test.ml.corrected
File "test.ml", line 12, characters 0-1:
 |open! Base
 |open! Stdio
 |
 |let get_href_hosts soup =
 |  Soup.select "a[href]" soup
 |  |> Soup.to_list
 |  |> List.map ~f:(Soup.R.attribute "href")
 |  |> List.filter_map ~f:(fun uri -> Uri.host (Uri.of_string uri))
 |  |> Set.of_list (module String)
 |
 |let%expect_test "" =
-|  print_endline "Whoa"
+|  print_endline "Whoa";
+|  [%expect {| Whoa |}]
```

But that's just using a different opam switch that kinda happens to work.  Is there a reasonable way of doing this?

**Comment by NathanReb** [comment]
I believe the issue here is that if you `cd` into your example directory and run dune from there it will try to find the root of the project using the `dune-project` file if I'm not mistaking.

That means that it won't be able to find the dependencies from the duniverse but will be able to pick them from opam if they happen to be available there.

I think the `data_only_dirs` we use to prevent `dune build` and `dune runtest` to also build and run examples play a role in this as well. I believe that this is to prevent dune from building examples that are purposefully erroneous. For those I'm not sure we can do much more than that unfortunately but for the other ones, we should be able to make them visible to dune which would allow you to use dependencies from the duniverse, either in MDX blocks are when you run the commands yourself.

**Comment by yminsky** [comment]
I'm missing some of the details above (what is data_only_dirs?), but do you have a view on the big-picture question?  i.e., is the next step to document how this is supposed to work in the current world, or to make some technical changes to make it easier? Or some combination?

**Comment by NathanReb** [comment]
I think it needs a bit of both.

We use dune's `data_only_dirs` stanza to mark `book/<chapter>/examples` directories as `data_only_dirs` so that dune doesn't not consider it's content and doesn't build anything there.

I think we need to make it so the examples are built and tested by dune and you can work on them seemlessly.

The erroneous examples are a different matter though as we have to "hide" them from dune since they won't build or the tests won't pass in those. They will have to be tested only through mdx (as are all examples atm). I think it's not much of an issue as we don't use too much of such examples and therefore the priority is to simplify the main use case.

I will try to remove the `data_only_dirs` and see how good of a workflow we can get for regular examples.

I think once we have a satisfying and documented setup for chapters and examples, I'll add a Makefile target or something to generate the base skeleton for new chapters.

Does that sound good?

**Comment by yminsky** [comment]
I'm not sure.  My temptation is to think that example directories should be standalone in more cases than just the case of a failing build.  Consider what a Dune chapter would look like.  We'd like to be free to set up whole little self-contained dune examples, with settings that differ from the outer build, with separate dune-project or dune-workspace files or whatever they're called these days.

Also, I think there are some issues if different examples define the same library, which is a common case when you're working through a sequence of examples, each one tweaked from the one before.

I think it would be good to at least have an easy mechanism for creating "standalone" examples that are in some sense independent.  Maybe that's what you're proposing here, except that you're suggesting using them only for the ones that error out, and I'm imagining using them at least somewhat more broadly.

**Comment by NathanReb** [comment]
You're right, I agree it's better to keep the examples self contained and isolated from each other.

If we want to keep things that way we won't be able to benefit from the repo's duniverse and therefore we won't be able to build them ourselves without an extra step to get their dependencies either through opam or in a local duniverse.
It also makes it hard to get the CI to check them automatically.

I think in a first iteration we'll have to stick to documenting how to setup examples directories and how to build and test them.

We could add `opam` files, `package` stanzas to their `dune-project` or even a `dune-get` (duniverse's lockfile) so that it's easier to figure out what an individual example's dependencies are and how to install them.
The downside to this is that it won't be checked anywhere and it could easily go out of sync.

In short I think there are two different concerns about examples:
1. How to setup examples in a way that's easy to work with, i.e. self contained dune-project with their own configuration that don't have to care about the outside world.
2. How to make them stand the test of time, i.e. ensure they always build and that their dependencies are well specified in the CI.

I'd say we almost have `1`, we can get there with a bit of documentation and a few opam files.

**Comment by yminsky** [comment]
Some notes from an off-github conversation:

So, I want to be able to write examples that behave essentially as standalone examples, working as they would if they were running in a fully independent Dune-structured project.

It should be hard to write those examples in way that silently fails.  Some of them are supposed to fail, and that should be possible, but the ideal would be that code that doesn't build should only be allowed when explicitly asked for.

It should also be easy to develop such examples, which means that you need a clear way to exercise these examples from the command line, in the same way that they'd be exercised to fill the blocks of the mdx files.

Here's a proposal for how to achieve the second property, that you can't easily fail by default.

For every embedded dune example, you could have a target specified somewhere whose failure is a build failure.  This could be default just be something like "dune build @all @test", or something along those lines. And you could override that when you wanted to.  (e.g., for when everything is broken, you could have the moral equivalent of running "true").

And, to say one more thing, I think it's essential that the CI test thse sub-projects automatically.  Otherwise, it's going to be impossible to maintain the correctness of these examples.
